amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/KinesisDataFetcher.java

```diff
@@ -120,7 +121,13 @@ public DataFetcherResult getRecords() {
 
         if (nextIterator != null) {
             try {
-                return new AdvancingResult(getRecords(nextIterator));
+                GetRecordsResponse getRecordsResponse = getRecords(nextIterator);
+                while (!isValidResponse(getRecordsResponse)) {
+                    log.error("{} : GetRecords response is not valid. nextShardIterator: {}. childShards: {}. Will retry GetRecords with the same nextIterator.",
+                              shardId, getRecordsResponse.nextShardIterator(), getRecordsResponse.childShards());
+                    getRecordsResponse = getRecords(nextIterator);
```

Comment:
Why do we need this extra getRecords call? Can't we just use the original call?

Reply:
The original idea is that if a response fails the validation, we should retry the same nextShardIterator. But implementing the retry in the while loop will cause throttling, so now moving the validation to getRecords method and throw RetryableRetrievalException if fails the validation.